Function pointer design (#282)
dotnet/designs: accepted/2022/FunctionPointers.md

# Function Pointers
#### Steve Harter
#### December 13, 2022

# Background
For additional context, see:
- The [original feature ask](https://github.com/dotnet/runtime/issues/11354)
- The [MetadataLoadContext feature ask](https://github.com/dotnet/runtime/issues/43791)
- The [7.0 introspection attempt](https://github.com/dotnet/runtime/pull/71516)
- [C# function pointer spec](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers)

> Function pointers, which have been in ECMA-335 and used by the CLR since the beginning were only recently supported by C# in C# v9.0 ad .NET v6. However, the .NET type system has not been updated to directly support them and they are currently exposed as the `IntPtr` type:

Reviewer: ```suggestion
Function pointers, which have been in ECMA-335 and used by the CLR since the beginning were only recently supported by C# in C# v9.0 and .NET v6. However, the .NET type system has not been updated to directly support them and they are currently exposed as the `IntPtr` type:
```

```cs
// using typeof:
Type type = typeof(delegate*<int, bool>); // System.IntPtr (currently)

// using Type.GetType():
delegate*<int, bool> fn = &MyFunctionPointer;
type = fn.GetType(); // System.IntPtr (currently)

static bool MyFunctionPointer(int n) { return n == 42; }
```

Reflection also returns `IntPtr` for both runtime-reflection through `FieldInfo\ParameterInfo\PropertyInfo` and static-reflection through `MetadataLoadContext`:
```cs
  FieldInfo fi = typeof(MyClass).GetField("_fn");
  Type type = fi.FieldType; // System.IntPtr (currently)

  unsafe class MyClass
  {
      public delegate*<int, bool> _fn; // A field as a function pointer
  }
```

> Treating function pointers as `IntPtr` essentially makes reflection introspection impossible, blocking scenarios that care about function pointers. The proposal is to change from `IntPtr` to `Type` which is a **breaking change**:

Reviewer 2: I have a scenario this spec would hopefully improve:

I want to create a method that takes a function pointer (any function pointer), and then reflects on the signature of that function pointer. Right now that's not possible, and the only way I could come up with to implement this was to add another parameter to the function that takes a MethodInfo for the target of the function pointer:

```cs
void DoSomething (void *fptr, MethodInfo method);
```

While it works, the calling code is quite ugly.

In order to take advantage of the new reflection capability, I guess I'd have to type the function pointer as object?

```cs
void DoSomething (object fptr)
{
    if (!ftpr.GetType ().IsFunctionPointer)
        throw new InvalidArgumentException ("Only function pointers!");
}
```

Unfortunately the calling code is not very nice on the eyes:

```cs

static void Y () { }

DoSomething (&Y); // doesn't work: error CS1503: Argument 1: cannot convert from '&method group' to 'object'
DoSomething ((delegate*<void>)&Y); // doesn't work either: error CS1503: Argument 1: cannot convert from 'delegate*<void>' to 'object'
DoSomething ((object)(delegate*<void>)&Y); // nope: error CS1503: Argument 1: cannot convert from 'delegate*<void>' to 'object'
DoSomething ((IntPtr)(delegate*<void>)&Y); // this works, but it's not very intuitive
DoSomething ((object)(IntPtr)(delegate*<void>)&Y); // this works, and it's also obvious why it works, but seriously, _3_ casts?
```

I thought about using generics, but that [doesn't work](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEAMACdBGAOgCUBXGAFyjAFMCBJC6gJwHsAHAZWYDcoBjagGcA3AFgAUBPQAmbAEE2bCQG8J2dbgDMuPADZcAFmwARFhxY1yAC1gBzADwAVAHzYAFI+wAzNuSYBKNQ1VcQ1sAF8g9XRtfH10IwBNd0DQ4KiIjJjsMkEIL2odeKMAWWgYd3xMAG0AXWwIJltBVLCQsLDTc0sbGAcEagAbalsIcmoAKnsE51c3ADJE/zE09UjxcKA):

```cs
void DoSomething<T> (T fptr) { }
void Y () { }
unsafe void F ()
{
    DoSomething<delegate*<void>> (&Y);
}
// error CS0306: The type 'delegate*<void>' may not be used as a type argument
```

Using a `System.Reflection.FunctionPointer` type as suggested at the end of this spec won't work (even if it were implemented), because there's no API to get the type of the function pointer.

What I'm hoping for is this to work somehow:

```cs
void DoSomething (FunctionPointer fptr) {}
static void Y () {}
void F ()
{
    DoSomething (&Y);
}
```

This would obviously require changes to both the runtime and the C# compiler, but it would be nice if the runtime part could be designed so that at least it's possible to implement the C# bits.

Author: > I want to create a method that takes a function pointer (any function pointer), and then reflects on the signature of that function pointer.

The polymorphic limitations here should be the same as dealing with a standard pointer - a pointer instance isn't boxable and thus you can't call `GetType()` on it:
```cs
int i = 1;
int* ip = &i;
ip.GetType(); // CS1061
```

Do you need to call the function pointer? If not, just passing the `Type` should be fine:

> void DoSomething (void *fptr, MethodInfo method);

You can modify this and just pass `Type` instead of `MethodInfo`:
```cs
static void Y() { }

delegate*<void> y = &Y;
DoSomething(y, y.GetType());
   
// Unless we want to call it, fnPtr is not required
void DoSomething(void* fnPtr, Type fnType)
{
    if (!fnType.IsFunctionPointer) throw new ArgumentException...
    
    // Inspect fnType here - parameter types, etc
}
```
Also in the "future apis" section there is a "MethodSignature" class that exposes the metadata from a function pointer and may also have late-bound Invoke capabilities. Not sure if that would help with your scenario.

Author: > I want to create a method that takes a function pointer (any function pointer), and then reflects on the signature of that function pointer.

The polymorphic limitations here should be the same as dealing with a standard pointer - a pointer instance isn't boxable and thus you can't call `GetType()` on it:
```cs
int i = 1;
int* ip = &i;
ip.GetType(); // CS1061
```

Do you need to call the function pointer? If not, just passing the `Type` should be fine:

> void DoSomething (void *fptr, MethodInfo method);

You can modify this and just pass `Type` instead of `MethodInfo`:
```cs
static void Y() { }

delegate*<void> y = &Y;
DoSomething(y, typeof(delegate*<void>));
   
// Unless we want to call it, fnPtr is not required
void DoSomething(void* fnPtr, Type fnType)
{
    if (!fnType.IsFunctionPointer) throw new ArgumentException...
    
    // Inspect fnType here - parameter types, etc
}
```
Also in the "future apis" section there is a "MethodSignature" class that exposes the metadata from a function pointer and may also have late-bound Invoke capabilities. Not sure if that would help with your scenario.

Reviewer 2: > You can modify this and just pass Type instead of MethodInfo:

Yes, that looks like it could be an improvement (avoids reflection at the callsite at least).

Thanks!

```cs
Type type = typeof(delegate*<int, bool>); // System.Type (proposed design)
bool isFunctionPointer = type.IsFunctionPointer; // A new method which returns 'true' here

> ```

Reviewer 3: Would it be possible to touch on the implications on castability? We're not just changing the reflection, but also how things are represented within the runtime.

This:

```csharp
unsafe
{
    object a = new delegate* unmanaged[MemberFunction]<void>[1];
    Console.WriteLine(a is delegate* unmanaged<void>[]);
    Console.WriteLine(a is delegate* unmanaged[SuppressGCTransition]<void>[]);
    Console.WriteLine(a is delegate* unmanaged[MemberFunction, SuppressGCTransition]<void>[]);

    var b = new delegate* unmanaged[MemberFunction]<void>[1];
    Console.WriteLine(b is delegate* unmanaged<void>[]);
    Console.WriteLine(b is delegate* unmanaged[SuppressGCTransition]<void>[]);
    Console.WriteLine(b is delegate* unmanaged[MemberFunction, SuppressGCTransition]<void>[]);
}
```

Currently prints:

```
True
True
True
False
False
False
```

I assume with this work we're going to make it print 6x False. (Notice how C# is trying to be clever but hitting a C#/runtime type system difference.)

For this one:

```csharp
unsafe
{
    object a = new delegate*<ref int, void>[1];
    Console.WriteLine(a is delegate*<out int, void>[]);
    Console.WriteLine(a is delegate*<in int, void>[]);

    var b = new delegate*<ref int, void>[1];
    Console.WriteLine(b is delegate*<out int, void>[]);
    Console.WriteLine(b is delegate*<in int, void>[]);
}
```

We'll keep the same output:

```
True
True
False
False
```

I.e. runtime type system will disagree with the C# type system, the same way how they disagree about e.g. array variance:

```csharp
Console.WriteLine(new int[1] is uint[]);
Console.WriteLine(((object)new int[1]) is uint[]);
```

Author: Thanks for the examples especially the array variance one.

I added a section called "Type Identity" - take a look at that especially now since the design doesn't include calling conventions for non-signature types but does include the invisible "CallKind".

The only custom modifiers exposed by the runtime are the built-in classes representing calling conventions - these classes are named `System.Runtime.InteropServices.CallConv*`:
- [CallConvCdecl](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvcdecl)
- [CallConvStdCall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvstdcall)
- [CallConvThiscall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvthiscall)
- [CallConvFastcall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvfastcall)
- [CallConvSuppressGCTransition](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvsuppressgctransition)
- [CallConvMemberFunction](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvmemberfunction)

C# in particular just maps calling conventions specified in a function pointer to `System.Runtime.InteropServices.CallConv*` types as metadata. These CallConv* types must live in the same assembly as defines `System.Object` (i.e. `System.Private.Corelib.dll`).

> 

Reviewer 4: > This would require a ECMA-335 change but that would be minimally breaking.

I'm not sure I agree with this. This would have required a change for every metadata reader to accept that this byte would be compressed. It would have meant that any tool that read in metadata on its own would have been broken when reading v6 assemblies that contained these new function pointers.

I'm not arguing there wasn't a better way, I am merely pushing back on this comment as I don't think the case was made it would have been "minimally breaking".

Reviewer 4: I just referenced ECMA-335 on this to ensure my criticism was founded in something rather than ignorance on details. I think this would have technically been an impactful breaking change but looking at the various implementations it might have been "okay". `calli` takes a token to a "stand-alone signature" - III.3.20. The "stand-alone signature" layout is defined in II.23.2.3 where the following is stated:

> The first byte of the Signature holds bits for HASTHIS, EXPLICITTHIS and calling 
convention – DEFAULT, VARARG, C, STDCALL, THISCALL, or FASTCALL. These are 
OR’d together

The assumption above being it is defined to be a "byte". This would imply to me that most parsers would simply read a byte and move on. However, the current implementation of the reader helpers for stand-along signatures do call `CorSigUncompressData()` so I guess we might have gotten away with it. Although it would have limited us to 32 different calling conventions and based on how we are now using base and modifying calling conventions, that might be a bit limiting.

Regardless of that, I appreciate the "off-topic note" as it did push me to look into this area.

Reviewer 5: For reference, https://github.com/dotnet/csharplang/pull/3441 is the main issue where the design was discussed.

> encode new calling conventions using custom modifiers

Roslyn used this encoding for other things before. IIRC, it was a no-brainer to follow the established patterns. This was not first use of custom modifier attached to return parameter to encode things that have nothing to do with return parameter. 

For example, compile and run the following:
```
Console.WriteLine(typeof(MyStruct).GetMethod("set_MyProperty").ReturnParameter.GetRequiredCustomModifiers()[0]);

public struct MyStruct
{
    public int MyProperty { get; init; }
}
```
You will see that `IsExternalInit` custom modifier attached to the return parameter.

> Although it would have limited us to 32 different calling conventions and based on how we are now using base and modifying calling conventions, that might be a bit limiting.

Right, there are hundreds of different calling conventions out there.

Also, one of the design goals of the encoding was that it should be possible to add new calling convention variants without modifying language or Roslyn compiler for each addition.

Reviewer 3: If we were to do a breaking change to the file format for this, I think it would be better to add a special calling convention that just says (on a high level): the next byte of the signature specifies a number N. After this, there's N tokens. That way we would still have the same flexibility, without misusing modopts. Of course that would be a structural file format breaking change, but not bigger than changing the meaning of the callconv byte to be a compressed integer. Either way, that ship has sailed.

Author: I brought this up since it was discussed in the 7.0 PR and thought continuing that discussion may be worthwhile.

Also FWIW I came across [these comments](https://github.com/dotnet/runtime/blob/4cd3e780c469e2416b6ebaba903dbea7c411a5b8/src/coreclr/vm/siginfo.cpp#L3742-L3746) that mentioned we used to read the calling conventions as a compressed integer.

> _Off-topic thoughts on current implementation_: using these CallConv* classes to represent calling conventions was added in V6 because new callling conventions needed to be added and would not fit within the 4 bits that were being used to encode the existing calling conventions based on ECMA-335. So the decision, arguably not the best, was made to encode new calling conventions using custom modifiers ince that would not affect existing metadata readers or require a ECMA-335 change. However, there are alternative approaches which may or may not have been considered -- one such approach would be to change the encompassing byte to be a 32-bit compressed integer encoding instead which .NET/CLI uses extensively elsewhere. This compressed integer approach is possible since the encompassing byte did not use bit 7 (`0x80`) which is the flag for a compressed integer to switch from one byte to two bytes. This would require a ECMA-335 change but that would be minimally breaking.

Reviewer 5: ```suggestion
_Off-topic thoughts on current implementation_: using these CallConv* classes to represent calling conventions was added in V6 because new callling conventions needed to be added and would not fit within the 4 bits that were being used to encode the existing calling conventions based on ECMA-335. So the decision, arguably not the best, was made to encode new calling conventions using custom modifiers since that would not affect existing metadata readers or require a ECMA-335 change. However, there are alternative approaches which may or may not have been considered -- one such approach would be to change the encompassing byte to be a 32-bit compressed integer encoding instead which .NET/CLI uses extensively elsewhere. This compressed integer approach is possible since the encompassing byte did not use bit 7 (`0x80`) which is the flag for a compressed integer to switch from one byte to two bytes. This would require a ECMA-335 change but that would be minimally breaking.
```

### Exposing custom modifiers on `FieldInfo`, `ParameterInfo` and `PropertyInfo`

> Although we don't want to expose custom modififiers from function pointer types that are not calling conventions, obtaining the other custom modifiers is useful for those who read metadata for varying reasons. The [`MetadataLoadContext`](https://learn.microsoft.com/dotnet/api/system.reflection.metadataloadcontext) class is the preferred approach to read metadata, since it supports reading any assembly without having to load the respective types into the runtime.

Reviewer 4: ```suggestion
Although we don't want to expose custom modifiers from function pointer types that are not calling conventions, obtaining the other custom modifiers is useful for those who read metadata for varying reasons. The [`MetadataLoadContext`](https://learn.microsoft.com/dotnet/api/system.reflection.metadataloadcontext) class is the preferred approach to read metadata, since it supports reading any assembly without having to load the respective types into the runtime.
```

"System.Int32()" | `delegate*<int>`
"System.Int32(System.String)" | `delegate*<string, int>`
"System.Int32(System.String, System.Boolean*&)" | `delegate*<string, ref bool*, int>`
"System.Int32()*" | `delegate*<int>*`
"System.Int32()[]" | `delegate*<int>[]`
"System.Int32()*[]" | `delegate*<int>*[]`
"System.Int32()()" | `delegate*<delegate*>int>>`
"System.Int32()[]" | `delegate*<int>[]`
"System.Boolean(System.String(System.Int32))" | `delegate*<delegate*<int, string>, bool>`

### `FullName` and `AssemblyQualifiedName` properties
Returns `null`. This is a simplification that avoids a [type name grammar update](https://docs.microsoft.com/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names) that would need to include grammar for custom modifiers as well as supporting  `Type.GetType()`, `Assembly.GetType()` and any other similar APIs that parse those strings in order to construct a function pointer from it.

Support for a grammar update is possible in the future, with a minimal breaking change here.

Note that this property is already nullable and does return `null` in open generic parameters cases as well.

### `Name` property

> Returns "`*()`".

Reviewer: Should this actually return `"*(" + comma_for_each_parameter + ")"` or `"*(" + parameter_count + ")"`? That seems more consistent with `Type.Name` in other scenarios such as multi-dimensional arrays and generic types.

Author: I do like the comma suggestion, but based on feedback below from Michal I think we should start with the empty string, and wait until we have a reason to change it to something more complicated.

> Returns "`*()`".

Author: I agree; changed to empty string.

The type hierarchy is covered somewhat [here](https://learn.microsoft.com/previous-versions/visualstudio/visual-studio-2008/2hf02550(v=vs.90)) and in ECMA\CLI: 
- Type
  - Value types
    - Built-in value types
    - User-defined value types
    - Enumerations
  - Reference types
    - **Pointer types**
    - Interface types
    - Self-describing types
      - Arrays
      - Class types
        - User-defined classes
        - Boxed value types
        - Delegates

# Proposed APIs
## System.Type
```diff
namespace System
{
    public abstract class Type
    {
        // These throw InvalidOperationException if IsFunctionPointer is false.
+       public virtual Type GetFunctionPointerReturnType();
+       public virtual Type[] GetFunctionPointerParameterTypes();
+       public virtual bool IsFunctionPointer { get; }

        // Required since this can't be determined by inspecting the results from GetFunctionPointerCallingConventions()
+       public virtual bool IsUnmanagedFunctionPointer { get; }

        // Exposing here without requiring a signature type is debatable; it is here since this

>         // information may be useful for runtime scenarios, and not just metadata inspection.

Reviewer 5: It is hard to tell whether it is more likely for runtime scenarios to need calling convention modopts as opposed other modopts.

Author: The design now says GetCallingConventions does not work unless it is for a signature type.

However, the existing implementation for CoreClr uses the "CallKind" in various locations, so that does introduce some equality concerns -- other compilers than C# may not use "CallKind" and instead may use mod opts.

> +       public virtual Type[] GetFunctionPointerCallingConventions(); 

Reviewer 3: Will this API synthesize the CallConv type for function pointers that have calling conventions that can be expressed the legacy way (e.g. `unmanaged[StdCall]` will not actually have the modopt in the signature).

Reviewer 4: Yes, I believe this should synthesize that. If we don't it becomes odd that if one provides `unmanaged[StdCall]`, there isn't a `modopt`, but if one provides `unmanaged[StdCall, SuppressGCTransitions]`, then one does appear. It might be more work, but I think ensuring we always project via attributes is important here.

    }
}
```

## System.Reflection.TypeDelegator
No new APIs, but new overrides to support delegation.
```diff
namespace System.Reflection
{
    public class TypeDelegator : TypeInfo
    {
+       public override Type GetFunctionPointerReturnType();
+       public override Type[] GetFunctionPointerParameterTypes();
+       public override bool IsFunctionPointer { get; }
+       public override bool IsUnmanagedFunctionPointer { get; }
+       public override Type[] GetFunctionPointerCallingConventions(); 
    }
}
```

## Signature types in MetadataLoadContext assembly
These extensions are located in the `MetadataLoadContext` assembly only.

> Instead of extensions we could add directly to the appropriate types (`Type` \ `FieldInfo` \ `PropertyInfo` \ `ParameterInfo`) as a hedge if we think we will want the runtime to implement these in the future. They would just throw an exception if used in the runtime for now. If we don't do that, and use `MetadataLoadContext` extensions as proposed, but then attempt to add support to the runtime later that will require a new set of extensions in the runtime assembly which may cause compiler-time ambiguity as to which extensions will be called.

Reviewer 5: Yes, I think it would be better to add them to the appropriate types.

Author: OK, I removed the extensions.

```diff
namespace System.Reflection
{
+   public static class MetadataLoadContextExtensions
+   {
        // These support function pointer types on FieldInfo.FieldType, PropertyInfo.PropertyType and ParameterInfo.ParameterType.

>         // InvalidOperationException is thrown when the corresponding type.IsFunctionPointer==false.

Reviewer 5: This would prevent retrieving function pointer details for composite types that include function pointer (array of function pointers, pointer to a function pointer, etc.). I think this method should work for any type.

Author: > I think this method should work for any type.

This does bring up some questions.  Composite types can be N-levels deep, forming a tree. So, for example, the root type may be an array of pointers to function pointers, but those function pointer types may have parameters which are also function pointers, etc. 

In this tree of types, are they all signature-types? Is it pruned to only support composite types? Also, signature-type equality would mean a non-signature-type and a signature-type are never equal.

If a signature-type is a class or struct, are its property and field types also signature-types and thus would support `Type.GetOptional\RequiredCustomModifiers()`? (at least for function pointer parameters, not necessarily MemberInfo parameters).

We could avoid all of this by having MetadataLoadContext always include custom modifiers for function pointer types. The type equality will be different from the runtime however.

Reviewer 5: Good questions! Here is one possible design: 

> In this tree of types, are they all signature-types?

Yes.

> Is it pruned to only support composite types?

No.

> Also, signature-type equality would mean a non-signature-type and a signature-type are never equal.

`UnderlyingSystemType` property of signature-type can return the underlying non-signature type. It would make signature-type and non-signature types equal.

> If a signature-type is a class or struct, are its property and field types also signature-types

It would require wrapping everything. It may get too complicated.

The simplest option would be to say that the signature types are meant just for dealing with signatures, make method/field/property inspection methods throw and tell people to use non-signature types to inspect the methods/fields/properties.

Author: Changed design to just assume signature types (field\property\parameter).

> +       public static Type GetSignatureType(this FieldInfo fieldInfo);

Reviewer 5: These names should match existing non-signature names. We have `FieldInfo.FieldType`, so the signature type getter should be `FieldInfo.GetSignatureFieldType()`

Author: Yep thanks

+       public static Type GetSignatureType(this PropertyInfo propertyInfo);
+       public static Type GetSignatureType(this ParameterInfo parameterInfo); 

>         // or "GetFunctionPointerSignatureType()?

Reviewer: `GetFunctionPointerSignatureType` might be a mouthful but I definitely think it's a more appropriate name, `GetSignatureType` isn't immediately clear about being for function pointers *only*. It's also more consistent with the naming of the other function pointer-related APIs introduced above.

Author: These are now "GetSignatureFieldType\GetSignaturePropertyType\GetSignatureParameterType".

I also added a comment that we don't want to add "FunctionPointer" prefix "just in case" there are other signature types we may want to consider.

        // These only work for types obtained from GetSignatureType().
        // An empty Type[] is returned for non-signature function pointer types.
        // InvalidOperationException is thrown when type.IsFunctionPointer==false.
+       public static Type[] GetRequiredCustomModifiers(this Type);
        // or "GetSignatureTypeRequiredCustomModifiers()?
+       public static Type[] GetOptionalCustomModifiers(this Type);
        // or "GetSignatureTypeOptionalCustomModifiers()?
+   }
}
```

Sample usage:
```c#
public class Holder
{
    public unsafe delegate*<in int, out int, void> _field;
}

FieldInfo fieldInfo = typeof(Holder).GetField("_field");

// Get the signature type from the field; this Type is like fieldInfo.FieldType but it includes custom mods
Type fnType = fieldInfo.GetSignatureType();

// Get the function pointer's first parameter's type

> Type[] p1Type = fnType.GetFunctionPointerParameterTypes()[0];

Reviewer: ```suggestion
Type p1Type = fnType.GetFunctionPointerParameterTypes()[0];
```

// Get the required modifiers from the signature type
Type[] p1ReqMods = p1Type.GetRequiredCustomModifiers();

// The required modifiers should include the 'in' modifier
Debug.Assert(p1ReqMods[0].GetType() == typeof(Runtime.InteropServices.InAttribute));
```

## System.Reflection.Emit.ILGenerator
See the existing [ILGenerator.EmitCalli()](https://docs.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli) for reference.
```diff
namespace System.Reflection.Emit
{
    public class ILGenerator
    {
+       public virtual void EmitCalli(
            // Must be OpCodes.Calli which matches existing EmitCalli convention
            // Not sure why this is redundant (for future Calli opcodes?)
+           System.Reflection.Emit.OpCodes opcode,
+           Type[] callingConventions,
+           Type? returnType,
+           Type[]? parameterTypes);

+       public virtual void EmitCalli(
+           System.Reflection.Emit.OpCodes opcode,
+           Type functionPointerType)
    }
}

> ```

Reviewer 5: Reflection.Emit will need some work to deal with function pointers correctly beyond the new EmitCalli overloads.

Consider this example that tries to emit a call to a method that takes function pointer https://gist.github.com/jkotas/79c7c673b7d892f41f6be5f16d380aed . It does work today. MethodInfo returns IntPtr instead of actual type, Reflection.Emit encodes method reference with a wrong signature, and we get MissingMethodException at runtime.

The problem is that this example cannot work even once this proposal is implemented. The example contains a function pointer with modreqs, but Reflection.Emit has no way to see the modreqs. It means that it is still going to emit reference with a wrong signature and we will get MissingMethodException at runtime.

This suggests that we cannot actually omit function pointer modopts/modreqs in runtime reflection object model if we want to have full support for function pointers in Reflection.Emit.

Reviewer 3: Wouldn't we run into the same problem if we wrap access to a `get; init;` setter like that? I think the problem is really that the runtime reflection system was not designed to introspect modifiers. It was designed to ignore them.

`Type` exposes various methods to construct pointers, arrays and generics through `Type.MakePointerType()` etc. However, there is no proposal for V8 to add similar methods to function pointers. The "MethodSignature" class would assist with that, and could either be passed as-is to represent function pointer metadata, or we could expose a `Type CreateFunctionPointerType()` method.

```diff
namespace System.Reflection
{
+   public sealed class MethodSignature
+   {
+       public MethodSignature(Type functionPointerType);
+       public MethodSignature(Type returnType, MethodParameter[]? parameters = null, Type[]? callingConventions = null);
+       public MethodParameter ReturnType { get; }
+       public MethodParameter[] Parameters { get; }
+       public Type[] CallingConventions { get; }
+       public bool IsUnmanaged { get; }

        // We could expose this mechanism if useful.
+       public Type CreateFunctionPointerType();

        // MethodInvoker class defined elsewhere
        // Throws InvalidOperation when used with MetadataLoadContext types
+       public MethodInvoker Invoker { get; }
+   }

> +   public sealed class MethodParameter

Reviewer 5: This looks like a subset of ParameterInfo.

Author: I added a comment that we might be able to use ParameterInfo. This was also somewhat discussed when we had "FunctionPointerParameterInfo" and we decided to just create a new type instead of re-using ParameterInfo.

### Exposing custom modifiers on `FieldInfo`, `ParameterInfo` and `PropertyInfo`
Although we don't want to expose custom modifiers from function pointer types that are not calling conventions, obtaining the other custom modifiers is useful for those who read metadata for varying reasons. The [`MetadataLoadContext`](https://learn.microsoft.com/dotnet/api/system.reflection.metadataloadcontext) class is the preferred approach to read metadata, since it supports reading any assembly without having to load the respective types into the runtime.

> The proposal is to expose functionality in the `MetadataLoadContext` assembly using type extensions that return "signature types" which are an extended function pointer type that has `GetRequiredCustomModifiers()` and `GetOptionalCustomModifiers()` methods which return all custom modifiers. These APIs only work with signature types that are returned from `FieldInfo`, `PropertyInfo` and `ParameterInfo` via `GetFunctionPointerSignatureType()`.

Reviewer 5: The reflection object model has a thing called signature type already: https://learn.microsoft.com/th-th/dotnet/api/system.type.issignaturetype

Are the signature types that we are talking about here same kind of signature types (ie `IsSignatureType` returns `true`) or are they something else? If they are something else, we should invent a new name for them.

Author: They are not the same -- I renamed them to "modified types".

For `IsSignatureType` see https://github.com/dotnet/runtime/issues/20377 which was added to make looking up generic methods from reflection easier. Also, I added `Type.CreateFunctionPointerType(Type[])` because lookup should also be possible based on a function pointer type.

    - User-defined value types
    - Enumerations
  - Reference types
    - **Pointer types**
    - Interface types
    - Self-describing types
      - Arrays
      - Class types
        - User-defined classes
        - Boxed value types
        - Delegates

# Proposed APIs
## System.Type
```diff
namespace System
{
    public abstract class Type
    {
        // These throw InvalidOperationException if IsFunctionPointer is false.
+       public virtual Type GetFunctionPointerReturnType();
+       public virtual Type[] GetFunctionPointerParameterTypes();
+       public virtual bool IsFunctionPointer { get; }

        // Required since this can't be determined by inspecting the results from GetFunctionPointerCallingConventions()
+       public virtual bool IsUnmanagedFunctionPointer { get; }

        // Exposing here without requiring a signature type is debatable; it is here since this
        // information may be useful for runtime scenarios, and not just metadata inspection.
+       public virtual Type[] GetFunctionPointerCallingConventions();

        // These only work for types obtained from GetFunctionPointerSignatureType().
        // An empty Type[] is returned otherwise.

> +       public Type[] GetRequiredCustomModifiers();
>         // or "GetSignatureTypeRequiredCustomModifiers()?
> +       public Type[] GetOptionalCustomModifiers();

Reviewer 5: ```suggestion
+       public virtual Type[] GetRequiredCustomModifiers();
        // or "GetSignatureTypeRequiredCustomModifiers()?
+       public virtual Type[] GetOptionalCustomModifiers();
```

>         // or "GetSignatureTypeOptionalCustomModifiers()?

Reviewer 5: `GetRequiredCustomModifiers` / `GetOptionalCustomModifiers` looks good to me.

Debug.Assert(type.IsFunctionPointer == true);
Debug.Assert(type.IsUnmanagedFunctionPointer == false);
Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));
Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
Debug.Assert(type.GetFunctionPointerCallingConventions().Length == 0);
```

```cs
// Simple unmanaged function pointer
Type type = typeof(delegate* unmanaged[Cdecl, MemberFunction]<int, bool>);
Debug.Assert(type.IsFunctionPointer == true);
Debug.Assert(type.IsUnmanagedFunctionPointer == true);
Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));

Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));

Debug.Assert(type.GetFunctionPointerCallingConventions().Length == 2);
Debug.Assert(type.GetFunctionPointerCallingConventions[0] == typeof(CallConvCdecl));
Debug.Assert(type.GetFunctionPointerCallingConventions[1] == typeof(CallConvMemberFunction));
```

## FieldInfo, PropertyInfo and ParameterInfo
```diff
namespace System.Reflection
{
    public class FieldInfo
    {

> +       public Type GetFunctionPointerSignatureType();
>     }
> 
>     public class PropertyInfo
>     {
> +       public Type GetFunctionPointerSignatureType();
>     }
> 
>     public class ParameterInfo
>     {
> +       public Type GetFunctionPointerSignatureType();

Reviewer 5: ```suggestion
+       public virtual Type GetFunctionPointerSignatureType();
    }

    public class PropertyInfo
    {
+       public virtual Type GetFunctionPointerSignatureType();
    }

    public class ParameterInfo
    {
+       public virtual Type GetFunctionPointerSignatureType();
```

Author: Thanks

+       public override bool IsFunctionPointer { get; }
+       public override bool IsUnmanagedFunctionPointer { get; }
+       public override Type[] GetFunctionPointerCallingConventions();
+       public override Type[] GetRequiredCustomModifiers();
+       public override Type[] GetOptionalCustomModifiers();
    }
}
```

## System.Reflection.Emit.ILGenerator
See the existing [ILGenerator.EmitCalli()](https://docs.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli) for reference.
```diff
namespace System.Reflection.Emit
{
    public class ILGenerator
    {
+       public virtual void EmitCalli(
            // Must be OpCodes.Calli which matches existing EmitCalli convention
            // Not sure why this is redundant (for future Calli opcodes?)
+           System.Reflection.Emit.OpCodes opcode,
+           Type[] callingConventions,
+           Type? returnType,
+           Type[]? parameterTypes);

+       public virtual void EmitCalli(
+           System.Reflection.Emit.OpCodes opcode,
+           Type functionPointerType)
    }
}
```

# Breaking changes

> The breaking change is: **`typeof(delegate*<...>)` and `Type.GetType()` will return a `Type` instance instead of `IntPtr`**

Reviewer 5: Observation: This breaking change is not actually strictly required to make the function pointers inspectable by reflection. We can keep the runtime current behavior and tell people to use signature types to gather any function pointer details.

(I am not trying to oppose the breaking change. Just sharing an observation.)

Author: Interesting; but I would say doing that goes against the goal of "correcting" the Type system.

This means that if we don't expose the calling conventions on `Type` for function pointers, some function pointer types will be castable and others not, and with no obvious way to inspect the `Type` to determine why. We could exposed just the "CallKind" byte on `Type`, but that is not desired since that will confuse and\or complicate the consumer's usage -- see "Normalization of calling conventions".

Since we must already do the work of exposing the calling conventions including the newer ones that are read via custom modifiers, it makes sense then to expose all custom modifieres (via the new `Type.GetRequiredCustomModifiers()` and `Type.GetOptionalCustomModifiers()`.

## Type identity 
### Using of void*

> A function pointer addresscan be cast to any function pointer type when using `void*`:

Reviewer 5: ```suggestion
A function pointer address can be cast to any function pointer type when using `void*`:
```

```cs
delegate*<string> fn_string = &StringMethod;
int i = ((delegate*<int>)(void*)fn_string)(); // Unsafe code allows any cast once void* is used
static string StringMethod() => "Forty-Two";
```

> And the IL just does a `ldftn` with no check:

Reviewer 5: Right, IL does not care about type identify of pointers or function pointers. There is one type identify for all pointers and function pointers at IL level. Any (function) pointer is implicitly convertible to any other (function) pointer at IL level.

Author: I re-worded that a bit so it doesn't sound like an issue.

```
Note the use of `ldtoken`:
```
ldtoken    method string *()
call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
ldtoken    method int32 *()
call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
call       bool [System.Runtime]System.Type::op_Equality(class [System.Runtime]System.Type, class [System.Runtime]System.Type)
```

### Casting
Using arrays to help:
```cs
object arrInt = new delegate*<int>[1];
var yikes = (delegate*<string>[])arrInt; // Expect InvalidCastException
```
```
.locals init (object V_0, method string *()[] V_1)
ldc.i4.1
newarr     method int32 *()
stloc.0
ldloc.0
castclass  method string *()[]
stloc.1
```
### `Is` keyword
```cs
object arrInt = new delegate*<int>[1];
bool f = arrInt is delegate*<string>[]; // Expect 'false'
```
IL:
```
.locals init (object V_0, bool V_1)
ldc.i4.1
newarr     method int32 *()
stloc.0
ldloc.0
isinst     method string *()[]
ldnull
cgt.un
stloc.1
```

> In all cases above* any calling conventions are included in type identity. For example,

Reviewer 5: We need to be consistent in what is included in the type identify. The other parts of the doc do not seem to be including calling conventions in the type identify. For example, the description of API shape:
```
        // These only work for types obtained from GetSignature*Type().
        // An empty Type[] is returned otherwise.
+       public virtual Type[] GetFunctionPointerCallingConventions();
+       public Type[] GetRequiredCustomModifiers();
+       public Type[] GetOptionalCustomModifiers();
```

I think it would be fine to omit the calling convention from the type identity (and thus do not consider the calling convention for runtime casting logic). When somebody is dealing with function pointers, it is unsafe code anyway. I do not think there is much value in the runtime to going extra mile to try to be type safe.

Author: The current design now says the calling conventions are not included in type identity however only for the newer mod opts. The "CallKind" I think still needs to be there, which introduces some inconsistencies. This needs to be discussed.

Author: I did remove the calling convention for type identity for unmodified types -- now it only tracks managed or not.

`delegate* unmanaged[Cdecl]<void>`
and
`delegate* unmanaged[FastCall]<void>`are not equal.

> _* not sure how the `is` (`isinst`) is implemented in the CLR and whether we can get the same behavior as `castclass`._

Reviewer 5: The casting logic behind `isinst`, `castclass` and `Type.IsAssignableFrom` is shared.

Author: Thanks - removed that comment. However, I did add a section here for cases where C# hard-codes the equality check instead of using `isinst` for cases where it "knows" what the result is.

Use the same behavior as standard pointers which is `IsValueType == false` and `IsClass == true`.

The type hierarchy is covered somewhat [here](https://learn.microsoft.com/previous-versions/visualstudio/visual-studio-2008/2hf02550(v=vs.90)) and in ECMA\CLI: 
- Type
  - Value types
    - Built-in value types
    - User-defined value types
    - Enumerations
  - Reference types
    - **Pointer types**
    - Interface types
    - Self-describing types
      - Arrays
      - Class types
        - User-defined classes
        - Boxed value types
        - Delegates

# Proposed APIs
## System.Type
```diff
namespace System
{
    public abstract class Type
    {
        // These throw InvalidOperationException if IsFunctionPointer is false.
+       public virtual Type GetFunctionPointerReturnType();
+       public virtual Type[] GetFunctionPointerParameterTypes();
+       public virtual bool IsFunctionPointer { get; }

+       public virtual bool IsUnmanagedFunctionPointer { get; }

        // These only work for types obtained from GetSignature*Type().
        // An empty Type[] is returned otherwise.
+       public virtual Type[] GetFunctionPointerCallingConventions();

> +       public Type[] GetRequiredCustomModifiers();
> +       public Type[] GetOptionalCustomModifiers();
>     }

Reviewer 5: ```suggestion
+       public virtual Type[] GetRequiredCustomModifiers();
+       public virtual Type[] GetOptionalCustomModifiers();
    }
```

Debug.Assert(type.GetFunctionPointerCallingConventions().Length == 0);
```

```cs
// Simple unmanaged function pointer
Type type = typeof(delegate* unmanaged[Cdecl, MemberFunction]<int, bool>);
Debug.Assert(type.IsFunctionPointer == true);
Debug.Assert(type.IsUnmanagedFunctionPointer == true);
Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));

Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));

Debug.Assert(type.GetFunctionPointerCallingConventions().Length == 2);
Debug.Assert(type.GetFunctionPointerCallingConventions[0] == typeof(CallConvCdecl));
Debug.Assert(type.GetFunctionPointerCallingConventions[1] == typeof(CallConvMemberFunction));
```

## GetSignature*Type() methods for `FieldInfo`, `PropertyInfo` and `ParameterInfo`
These return a function pointer `Type` that has references to custom modifiers.

If not a function pointer, these methods return the value from the appropriate `FieldInfo.FieldType`, `PropertyInfo.PropertyType` or `ParameterInfo.ParameterType` property.

```diff
namespace System.Reflection
{

>     public class FieldInfo
>     {
> +       public Type GetSignatureFieldType();
>     }
> 
>     public class PropertyInfo
>     {
> +       public Type GetSignaturePropertyType();
>     }
> 
>     public class ParameterInfo
>     {
> +       public Type GetSignatureParameterType();
>     }

Reviewer 5: ```suggestion
    public abstract class FieldInfo
    {
+       public virtual Type GetSignatureFieldType();
    }

    public abstract class PropertyInfo
    {
+       public virtual Type GetSignaturePropertyType();
    }

    public abstract class ParameterInfo
    {
+       public virtual Type GetSignatureParameterType();
    }
```

Reviewer 5: Also, what is going to be the default implementation of these methods?

Author: The default implementation will just return `FieldInfo.FieldType`, etc. I don't think we want to throw if a non function pointer.

Author: Update: I changed these to return `null` if the FieldType etc is not a function pointer. Also related is that a modified type is just a "wrapper" on top of the underlying type that doesn't pretend to be the actual underlying type.

C# in particular just maps calling conventions specified in a function pointer to `System.Runtime.InteropServices.CallConv*` types as metadata. These CallConv* types must live in the same assembly as defines `System.Object` (i.e. `System.Private.Corelib.dll`).

### Normalization of calling conventions
Calling conventions can be specified in one of two ways in metadata:
- The older "CallKind" enum (which has issues with not being large enough to hold the newer calling conventions).
- The newer approach of using modopts on the return parameter.

In ECMA, "CallKind" is used with "CallConv":"
```
callConv :'instance' callConv 
 | 'explicit' callConv 
 | callKind 
 ; 
callKind : /* EMPTY */ 
 | 'default' 
 | 'vararg' 
 | 'unmanaged' 'cdecl' 
 | 'unmanaged' 'stdcall' 
 | 'unmanaged' 'thiscall' 
 | 'unmanaged' 'fastcall' 
```
and in exposed in various enums and types. Here's the raw values:
```
CALLCONV_DEFAULT       = 0x0,  

CALLCONV_VARARG        = 0x5,  
CALLCONV_FIELD         = 0x6,  
CALLCONV_LOCAL_SIG     = 0x7,  
CALLCONV_PROPERTY      = 0x8,  
CALLCONV_UNMGD         = 0x9,  
CALLCONV_GENERICINST   = 0xa,  

> CALLCONV_NATIVEVARARG  = 0xb,  

Author: I removed that section.

CALLCONV_MAX           = 0xc,  

CALLCONV_MASK          = 0x0f,  
CALLCONV_HASTHIS       = 0x20,  
CALLCONV_EXPLICITTHIS  = 0x40,  
CALLCONV_GENERIC       = 0x10  
```

The calling convention is considered managed if the callKind is "default" or "vararg", so the proposed `Type.IsUnmanagedFunctionPointer` does not require looking at the newer modopts (at least currently; it is possible new managed calling conventions are introduced at some point).

**Open issue: currently this design does not provide a way to determine if the managed calling convention is "default" vs "vararg" or "instance" vs. "explicit" (for the "this" parameter to instance methods).** However, we could easily add new methods to expose this (e.g. using two `bool` properties `IsExplicitThis` and `IsVarArg` or add a future-extensible implementation with enums instead of `bool`).

> C# only uses the newer modopts approach when necessary; see the [c# spec section](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers#metadata-representation-of-calling-conventions). However, we should assume any valid encoding.

Reviewer 5: > C# only uses the newer modopts approach when necessary

All IL emitters are required to do this to allow assemblies produced by different compilers to interoperate.

`SignatureCallingConvention.Unmanaged` + `CDecl` modopt is invalid encoding.

Author: Thanks. Updated

> When returning the calling conventions through `Type.GetFunctionPointerCallingConventions()` any calling convention encoded using "CallKind" will return the appropriate CallConv* type, just like it was added through a modopt. This normalization avoids having to expose the "CallKind" in the API and makes it simpler for callers since there is just a single method to return the calling conventions no matter how they were encoded. The `Type.IsUnmanagedFunctionPointer` property was added since there are not CallConv* types for mananged that would otherwise have helped determine this.

Reviewer 5: ```suggestion
When returning the calling conventions through `Type.GetFunctionPointerCallingConventions()` any calling convention encoded using "CallKind" will return the appropriate CallConv* type, just like it was added through a modopt. This normalization avoids having to expose the "CallKind" in the API and makes it simpler for callers since there is just a single method to return the calling conventions no matter how they were encoded. The `Type.IsUnmanagedFunctionPointer` property was added since there are not CallConv* types for managed that would otherwise have helped determine this.
```

The reason to support calling conventions through `Type.GetFunctionPointerCallingConventions()` with the runtime are that the calling conventions may be useful in runtime-cases (not just introspection cases) and because there is a larger design factor with the CoreClr: the internal function pointer implementation uses the "CallKind" byte to hold the legacy calling conventions for things like determining if the VARARG convention is used. In addition, the CoreClr has a concept of "TypeKey" and "TypeHandle" and they are 1:1 today meaning that all `Type` instances are fully self-describing and thus, we can't have additional state hanging off of `Type` that is not part of the "key", such as calling conventions -- thus also the reason for bifurcation of signature types and non-signature types.

So we have two choices for exposing the calling conventions:

> 1) Expose only for only signature types.

Author: Yes thanks I did that.

Function pointers, which have been in ECMA-335 and used by the CLR since the beginning were only recently supported by C# in C# v9 and .NET v6. However, the .NET type system has not been updated to directly support them and they are currently exposed as the `IntPtr` type:
```cs
Type type = typeof(delegate*<int, bool>); // System.IntPtr (currently)
```

Reflection also returns `IntPtr` for both runtime-reflection through `FieldInfo\ParameterInfo\PropertyInfo` and static-reflection through `MetadataLoadContext`:
```cs
  FieldInfo fi = typeof(MyClass).GetField("_fn");
  Type type = fi.FieldType; // System.IntPtr (currently)

  unsafe class MyClass
  {
      public delegate*<int, bool> _fn; // A field as a function pointer
  }
```

Treating function pointers as `IntPtr` essentially makes reflection introspection impossible, blocking scenarios that care about function pointers. The proposal is to change from `IntPtr` to `Type` which is a **breaking change**:
```cs
Type type = typeof(delegate*<int, bool>); // System.Type (proposed design)
bool isFunctionPointer = type.IsFunctionPointer; // A new method which returns 'true' here
```

> Note that Mono returns "System.MonoFNPtrFakeClass" for the type name, not "IntPtr". This design applies to Mono as well and is expected that both runtimes are consistent going forward.

Reviewer 4: @lambdageek Any concerns with the above? Looking across github I see relatively few mentions of `MonoFNPtrFakeClass` so assume this is likely to be uninteresting.

Reviewer 6: I _think_ `MonoFNPtrFakeClass` is just used to fill in `MonoClass:name` (because we often try to use that field in debug output and the like and null-checking it everywhere is probably a pain).  That it happens to also be used when you call `type.FullName` or `type.ToString()` is incidental and we could change it.  Internally in mono function pointer types are quite distinct from other types - having new behavior for them shouldn't be an architectural problem (implementation might be another matter: there have been - and probably still are - many `switch` statements in the C code that don't have a good fn ptr branch)

@AaronRobinsonMSFT  this intro section is confusing given what comes later in the design.
You're not proposing `typeof(delegate*<int,bool>) == typeof(System.Type)` . The behavior today is that `typeof(delegate*<int,bool) == typeof(System.IntPtr)`, and what you want is an actual new instance of `System.Type` for each function pointer type, right?

Author: >  and what you want is an actual new instance of System.Type for each function pointer type, right?

Yes a separate instance for each unique function pointer (ignoring custom modifiers). I updated this intro section. Thanks

## Modified types
Obtaining all custom modifiers including `ref\out\in` is useful for those who read metadata for varying reasons. The proposal to support that is to expose functionality that returns a "modified type" which, for function pointers, is a wrapper over the unmodified function pointer and intercepts calls to `GetFunctionPointerParameters()` to return a "modified type" for each parameter. From that parameter modified type, `GetRequiredCustomModifiers()` and `GetOptionalCustomModifiers()` can be used.

> A modified type always start with a "root" function pointer obtained from `FieldInfo.GetModifiedFieldType()`, `PropertyInfo.GetModifiedPropertyType()` and `ParameterInfo.GetModifiedParameterType()`. **discussion topic**: a modified type "tree" ends with the modified parameter types. For example, if a field is a function pointer that has parameters and those parameters are also function pointers, the secondary function pointers are not accessible as modified types. This limitation was added to simplify the design. Without it, an N-level tree of modified types would need to be created, recusively. It is possible to add this in the future, however.

Reviewer 5: > This limitation was added to simplify the design.

Does it simplify the design, or just the implementation? I think it is just the latter, and it is pretty marginal simplification.

Reviewer 5: > It is possible to add this in the future, however.

Is it going to be a breaking change to add it later?

Author: Both the design and implementation. If we don't support "nested" types, the implementation becomes trivial -- implemented in managed code only, using the existing "signature" field already there for FieldInfo, PropertyInfo and MethodInfo.

If we supported "nested" types, then the [discussion here](https://github.com/dotnet/designs/pull/282#discussion_r1049031775) becomes relevant. Do we want to support obtaining custom mods from all of the following function pointers?
```
// Assumed
delegate*<int> s_0 = null;

// Arrays?
delegate*<int>[] a_0 = new delegate*<int>[] { };
delegate*<int>[][] a_1 = null!;
delegate*<int>[,] a_2 = null!;
// etc.

// Pointers?
delegate*<int>* p_0 = null;
delegate*<int>** p_1 = null;
delegate*<int>*** p_2 = null;
// etc.

// Arrays + pointers?
delegate*<int>*[] ap_0 = null!;
delegate*<int>**[] ap_1 = null!;
delegate*<int>*[][] ap_2 = null!;
// etc.

// Nested fnptrs?
delegate*<delegate*<int>> fp_0 = null;
// with pointers
delegate*<delegate*<int>*> fp_1 = null;
// with arrays
delegate*<delegate*<int>[]> f2_3 = null;
// with arrays + pointers
delegate*<delegate*<int>*[]> f2_5 = null;
// etc...
```

Author: I think we should try to support the above, but need to understand how these nested\inner types are populated -- eager vs. lazy for example. If eager, and in the same "loop", then this should be trivial to support.

Reviewer 5: It should be fine to start with eager population for simplicity. (We can always switch to lazy if the eager population of the full modified type shows up as a performance issue.)

Author: Resolving; the current PR supports the above in a lazy-load manner. It also supports generics such as `Tuple<delegate*<void>[]` which is allowed by the compiler since the function pointer is not directly the generic signature type.

> A modified type always start with a "root" function pointer obtained from `FieldInfo.GetModifiedFieldType()`, `PropertyInfo.GetModifiedPropertyType()` and `ParameterInfo.GetModifiedParameterType()`. **discussion topic**: a modified type "tree" ends with the modified parameter types. For example, if a field is a function pointer that has parameters and those parameters are also function pointers, the secondary function pointers are not accessible as modified types. This limitation was added to simplify the design. Without it, an N-level tree of modified types would need to be created, recusively. It is possible to add this in the future, however.

Reviewer 5: It sounds like you are also proposing to cut full inspection of things like arrays of function pointers in the initial implementation. Is that correct? Is it going to be a breaking change to add it later?

## Supporting both runtime- and and static-reflection
Previous discussions included the suggestion that both calling conventions and modified types are only exposed from the [`MetadataLoadContext`](https://learn.microsoft.com/dotnet/api/system.reflection.metadataloadcontext) class since it is the preferred approach to read metadata (vs. runtime reflection) due to being agnostic to the runtime referenced by the various reflected types. However, this design proposes full fidelity between the runtime and `MetadataLoadContext` (other runtimes besides the CoreCLR, including Mono and AOT runtimes, will need to be updated of course).

Note that the `Type.IsUnmanagedFunctionPointer` property works for both modified and unmodified types.

> For This design assumes (1) - expose only for modified types even though there are some equality issues.

Reviewer 5: > For This

Unfinished edit?

```cs
Console.WriteLine(new int[1] is uint[]); // false
Console.WriteLine(((object)new int[1]) is uint[]); // true
```

### With modifiers + unmodified types
The above examples only compare parameter types without calling conventions or modifiers.

An unmodified type does include managed vs. unmanaged as part of its type identity:
```cs
object o = new delegate*<int>[1];
Console.WriteLine(o is delegate*<int>[]); // true
Console.WriteLine(o is delegate* unmanaged[Cdecl]<int>[]); // false

var fn = new delegate*<int>[1];
Console.WriteLine(fn is delegate*<int>[]); // true
Console.WriteLine(fn is delegate* unmanaged[Cdecl] <int>[]); // false
```

but unmanaged calling conventions are not part of the type identity:
```cs
object o1 = new delegate* unmanaged[SuppressGCTransition]<int>[1];
Console.WriteLine(o1 is delegate* unmanaged[MemberFunction]<int>[]); // true

object o2 = new delegate* unmanaged[CDecl]<int>[1];
Console.WriteLine(o2 is delegate* unmanaged[StdCall]<int>[]); // true
```

In addition, all non-CallConv* modifiers, such as `ref\in\out\const` are not considered part of the type identity either.

### With modified types

> A modified type is a simple wrapper that does not forward the other various `Type` members to the underlying type. It does, however intercept `GetFunctionPointerParameters()` to return modified type parameters. It also implements `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()` which throw on unmodified types.

Reviewer 3: I'd consider `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()` to return `null` if it's not the special modified type. It matches how e.g. `GetElementType()` returns null on non-arrays.

Reviewer 5: Returning `null` would make it different from the existing `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()` methods on `PropertyInfo`, `FieldInfo` and `MethodInfo`. I think it would be better to throw instead of returning null. People can check `IsModifiedType` property upfront if they are not sure whether they are dealing with modified type.

Reviewer 3: The existing GetOptionalCustomModifiers on PropertyInfo, FieldInfo and MethodInfo return an empty array if there's no modifiers. Should we match that one?

Reviewer 4: > The existing GetOptionalCustomModifiers on PropertyInfo, FieldInfo and MethodInfo return an empty array if there's no modifiers. Should we match that one?

I would prefer an empty array rather than `null`.

Author: > The existing GetOptionalCustomModifiers on PropertyInfo, FieldInfo and MethodInfo return an empty array if there's no modifiers. Should we match that one?

Yeah I think so. Throwing however may help users catch bugs (exception would say "only works on modified types"). Returning empty is preferred over `null` for consistency. So at this point I have a slight preference for returning empty for consistency vs. throwing.

Author: Any opinion on not whether we should try to forward all Type members (except for `Get*CustomModifiers` and `IsModifiedType`) to the underlying type to try to make it act like the underlying type? If we do that, then we probably should cache modified Type instances so the same instance is returned when everything matches including the custom mods. Caching however adds global memory for things that may only be used once and also may hamper assembly unloadability...

An option lightly discussed prior is to expose a non-Type instance like "SignatureMethod" mentioned in the design, but abandoned for various reasons. However, another approach is returning a wrapper for the custom mods like a "TypeModifiers" class:
```cs
public sealed class TypeModifiers
{
    public Type UnderlyingType { get; } // For access to fnptr parameter types etc.
    public Type[] GetOptionalCustomModifiers();
    public Type[] GetRequiredCustomModifiers();
}

// and returned by FieldInfo etc instead of System.Type
public abstract class FieldInfo
{
    public virtual Type? GetFieldTypeModifiers();
}
```
which makes the API more understandable IMO. i.e. no potential misuse of `Type.Get*CustomModifiers()`.

Thus there isn't a `Type` identity for a modified type since it is created lazily and not cached. Comparing using `==` or `Equals()` will return `false`.

```cs
public class Holder
{
    public unsafe delegate* unmanaged[SuppressGCTransition]<int> _field;
}

Type modifiedType = typeof(Holder).GetField("_field").GetModifiedFieldType()!;
Console.WriteLine(modifiedType.IsModifiedType); // true
Console.WriteLine(modifiedType.UnderlyingSystemType); // (function pointer)

> Console.WriteLine(modifiedType.IsFunctionPointer); // false

Reviewer 3: Why not have it return `true`?

The modified types could just derive from `TypeDelegator` (either conceptually, or actually derive from the type) and delegate everything to the underlying type except for the questions about modifiers.

Console.WriteLine(modifiedType.UnderlyingSystemType.IsFunctionPointer); // true

Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions().Length); // 2
Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[0]); // CallConvCdecl
Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[1]); // CallConvMemberFunction

Type modifiedType2 = typeof(Holder).GetField("_field").GetModifiedFieldType()!; // same field...

> Console.WriteLine(modifiedType == modifiedType2); // false (a different type instance; not cached)

Reviewer 3: Are we going to do extra work to make this not work?

`==` is implemented by calling `Equals` and unless we override it, the comparison would be done on `UnderlyingSystemType` and return `true`. I think that might be the desired behavior. It is still the same type, just with extra metadata attached to it.

Reviewer 3: (Or maybe you meant `object.ReferenceEquals` instead of `==`?)

Author: Based on other feedback as well around the `GetModified*Type()` methods returning `null` or not, I think it makes the most sense to treat the modified types as essentially the underlying type except with different identity (as expected, since one will have custom modifiers).

### `Name`, `FullName`, `AssemblyQualifiedName`, `Assembly` and `Namespace` properties
A modified type is an internal class called "System.RuntimeModifiedType" class that derives from a "System.ModifiedType" which derives from "System.Type". So these methods just return the expected values there.

### `GetFunctionPointerParameters()`
This returns modified types for each function pointer parameter which allows the user to call `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()`.

This is listed here since this method also works on unmodified types, although unmodified types don't allow `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()` to be called (an `InvalidOperationException` is thrown).

# Proposed APIs
## System.Type
```diff
namespace System
{
    public abstract class Type
    {
        // Throws InvalidOperationException if IsFunctionPointer = false.
+       public virtual Type GetFunctionPointerReturnType();
+       public virtual Type[] GetFunctionPointerParameterTypes();
+       public virtual bool IsFunctionPointer { get; }

> +       public virtual bool IsUnmanagedFunctionPointer { get; }

Author: Should this just be `IsManagedFunctionPointer`? (remove the "un")

Reviewer 5: > Should this just be `IsManagedFunctionPointer`? (remove the "un")

I think it should be "unmanaged" for consistency with existing names. Unmanaged is used in all existing related names: `unmanaged` keyword in C#, `UnmanagedCallersOnlyAttribute`, `SignatureCallingConvention.Unmanaged`.

        // Returns whether the type is a modified type.
        // A modified type is either:
        //  - A function pointer obtained though FieldInfo\PropertyInfo\ParametersInfo.GetModified*Type().
        //  - A function pointer parameter obtained by 'GetFunctionPointerParameterTypes()[index]' from a modified function pointer type.
+       public virtual bool IsModifiedType { get; }

        // Updated semantics: if IsModifiedType = true, returns the unmodified type such as `typeof(Int32)`.
        public virtual bool UnderlyingSystemType { get; }

        // These only work if IsModifiedType = true. InvalidOperationException otherwise.
+       public virtual Type[] GetRequiredCustomModifiers();
+       public virtual Type[] GetOptionalCustomModifiers();
        // Throws InvalidOperationException if UnderlyingSystemType.IsFunctionPointer = false
+       public virtual Type[] GetFunctionPointerCallingConventions(); 

>         // This is used with 'Type.GetMethod(..., Type[] types)' to find a method with parameters that

Reviewer 5: You can overload based on calling convention. For example, the following compiles fine:
```
static void foo(delegate* unmanaged[Stdcall]<void> f) { }
static void foo(delegate* unmanaged[Cdecl]<void> f) { }
```
Is this going to find the first overload at random? Or are going to allow passing modified types to `GetMethod` calls?

Author: The right thing to do is to throw `AmbiguousMatchException` and then, I suppose, add a `MakeModifiedFunctionPointer()` overload that takes the extra custom mods.

Adding such an overload would be ugly (2 dimension jagged array?) unless we add a helper class, like the "MethodSignature" proposed in the design.

After thinking about this, I may move this feature to the "future" section. For now, to avoid a breaking change later, we throw `NotSupportedException` if there is a function pointer type in the `Type[]` passed to `Type.GetMethod()`. The workaround is the user would need to call `GetMethods()` and then manually loop through them and inspect the custom modifiers on the parameters themselves.

Author: (design updated for this)

        // are function pointers. For a similar reference, see Type.MakePointerType().

> +       public static Type MakeFunctionPointerType(Type[] functionPointerArguments);

Reviewer 3: I'd suggest separating the return type - it's a trial and error to remember whether return type comes first or last in schemes where it's a single array.

Do we also need a bool to signify if this is an unmanaged/managed function pointer?

(My reading of this is that `MakeFunctionPointerType` would return a type that has a valid `TypeHandle` and these would be the only constituents that form the runtime identity .)

Author: Thanks. I updated the design for separate parameters + isUnmanaged but based on other feedback, moved this to the "future" section. The work-around is not so bad, plus likely to be very rare anyway unlike the original support which was added for looking up generic methods and thus much more likely to be used.

> +       public static Type MakeFunctionPointerType(Type[] functionPointerArguments);

Author: Sorry the "similar reference" should be [`Type.MakeGenericSignatureType`](https://learn.microsoft.com/dotnet/api/system.type.makegenericsignaturetype).

From a modified parameter type, the custom modifiers can be accessed via `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()`.

```diff
namespace System.Reflection
{
    public abstract class FieldInfo
    {
+       public virtual Type? GetModifiedFieldType() => null;
        // or "GetFieldModifiedType"?
    }

    public abstract class PropertyInfo
    {
+       public virtual Type? GetModifiedPropertyType() => null;
    }

    public abstract class ParameterInfo
    {
+       public virtual Type? GetModifiedParameterType() => null;
    }
}
```

### Examples
```c#
public class Holder
{
    public unsafe delegate*<in int, out int, void> _field;
}

FieldInfo fieldInfo = typeof(Holder).GetField("_field");

// Get the modified type from the field; this Type is like fieldInfo.FieldType but it includes custom mods

> Type fnType = fieldInfo.GetModifiedFieldType();

Reviewer 5: ```suggestion
Type fnType = fieldInfo.GetModifiedFieldType()!;
```
What is the right way to deal with `GetModifiedFieldType` possibly returning null? Add `!` everywhere?

Author: I updated the design to return non-null. By doing this, we may return an unmodified type. The design was updated to state that it really shouldn't matter to the caller since an unmodified and modified type have the same semantics for things like GetRequiredCustomModifiers() where we just return an empty array (and don't throw for example).

Author: To keep consistent semantics, `GetModified*Type()` will always return a "modified type" even when it could return a "runtime type". This is because a modified type (for the runtime anyway) will not derive from the internal `RuntimeType` and instead derive from `TypeDelegator` which does not derive from `RuntimeType` and we want to ensure the same semantics (throw or allow) for the methods that only take `RuntimeType` (performed when validating parameters).

For example, if we don't "need to" create a modified type in `GetModified*Type()` (because there are no modifiers in that type or any nested types), and instead create a runtime type (which I'm proposing we DON'T want to do), the user may start passing those to methods that only take runtime types which will work fine until  `GetModified*Type()` needs to return a modified type which fails the "pit of success" test.

## Custom modifiers
The [7.0 introspection attempt](https://github.com/dotnet/runtime/pull/71516) returned all custom modifiers including modifiers not used or needed by the runtime, such as parameter modifiers for `ref`, `out` and `in`. Due to concerns with that approach, the function pointer feature was pulled from late in 7.0 in order to properly vet and approve a new design (this design) for v8. The intent is to add this functionality early in v8 which will also allow time for feedback.

Consider the following where `ref\out\in` all map to a reference (`&`) by the runtime:
```cs
class MethodHolder
{
    public static void M1(in int i) { }
    public static void M2(ref int i) { }
    public static void M3(out int i) { i = 42; }
}

Type t1 = typeof(MethodHolder).GetMethod("M1").GetParameters()[0].ParameterType;
Type t2 = typeof(MethodHolder).GetMethod("M2").GetParameters()[0].ParameterType;
Type t3 = typeof(MethodHolder).GetMethod("M3").GetParameters()[0].ParameterType;

Debug.Assert(t1 == t2 && t2 == t3); // True today; all are "int reference type" (int&)
```

> Also consider [ILGenerator.EmitCalli()](https://docs.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli(system-reflection-emit-opcode-system-runtime-interopservices-callingconvention-system-type-system-type())) as a reference implementation which does not support modifiers:

Reviewer 5: Nit: Fix all links to use `learn.microsoft.com` instead of `docs.microsoft.com`?

```cs
object o1 = new delegate* unmanaged[SuppressGCTransition]<int>[1];
Console.WriteLine(o1 is delegate* unmanaged[MemberFunction]<int>[]); // true

object o2 = new delegate* unmanaged[CDecl]<int>[1];
Console.WriteLine(o2 is delegate* unmanaged[StdCall]<int>[]); // true
```

In addition, all non-CallConv* modifiers, such as `ref\in\out\const` are not considered part of the type identity either.

### With modified types
A modified type acts like the underlying type with the exception that it overrides `GetFunctionPointerParameters()` to return the modified parameter types. It has an identity separate than the underlying types.

```cs
public class Holder
{
    public unsafe delegate* unmanaged[SuppressGCTransition]<int> _field;
}

Type modifiedType = typeof(Holder).GetField("_field").GetModifiedFieldType();
Console.WriteLine(modifiedType.IsModifiedType); // true
Console.WriteLine(modifiedType.UnderlyingSystemType); // (unmodified function pointer)
Console.WriteLine(modifiedType.IsFunctionPointer); // false
Console.WriteLine(modifiedType.UnderlyingSystemType.IsFunctionPointer); // true

Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions().Length); // 2

> Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[0]); // CallConvCdecl

Reviewer 5: Where did `CallConvCdecl` came from? The function pointer above uses default calling convention, it does not use Cdecl.

Author: Oops, typos caused by refactoring that section earlier. Updated.

Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[0].IsModifiedType); // true
Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[1]); // CallConvMemberFunction

Type modifiedType2 = typeof(Holder).GetField("_field").GetModifiedFieldType(); // same field...

> Console.WriteLine(modifiedType == modifiedType2); // true (same instance)

Reviewer 5: ```suggestion
Console.WriteLine(modifiedType == modifiedType2); // true (types are equal)
Console.WriteLine(modifiedType.Equals(modifiedType.UnderlyingSystemType)); // true (modified and underlying types are equal)
Console.WriteLine(ReferenceEquals(modifiedType, modifiedType2)); // undefined. The implementation is not required to return same instance.
```

Operator `==` is overloaded for System.Type, so `==` does not check whether it is the same object instance.

Same instance check would be `Console.WriteLine(ReferenceEquals(modifiedType, modifiedType2))`. I do not think we need to guarantee that the implementation returns same instance.

Author: Sure thanks. I see that `==` is overridden and calls Equals which [inspects UnderlyingSystemType](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Type.cs#L636)

---|---
"System.Void()" | `delegate*<void>`
"System.Int32()" | `delegate*<int>`
"System.Int32(System.String)" | `delegate*<string, int>`
"System.Int32(System.String, System.Boolean*&)" | `delegate*<string, ref bool*, int>`
"System.Int32()*" | `delegate*<int>*`
"System.Int32()[]" | `delegate*<int>[]`
"System.Int32()*[]" | `delegate*<int>*[]`
"System.Int32()()" | `delegate*<delegate*<int>>`
"System.Boolean(System.String(System.Int32))" | `delegate*<delegate*<int, string>, bool>`

### `FullName` and `AssemblyQualifiedName` properties
Returns `null`. This is a simplification that avoids a [type name grammar update](https://docs.microsoft.com/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names) that would need to include grammar for custom modifiers as well as supporting  `Type.GetType()`, `Assembly.GetType()` and any other similar APIs that parse those strings in order to construct a function pointer from it.

Support for a grammar update is possible in the future, with a minimal breaking change here.

Note that this property is already nullable and does return `null` in open generic parameters cases as well.

### `Name` property

> Returns `String.Empty()`. Any other syntax here is somewhat arbitrary, and since the property is not nullable, the empty string seems fine unless a compelling reason is found. 

Reviewer 5: ```suggestion
Returns `string.Empty`. Any other syntax here is somewhat arbitrary, and since the property is not nullable, the empty string seems fine unless a compelling reason is found. 
```

+       public virtual Type[] GetOptionalCustomModifiers();

        // Throws InvalidOperationException if IsFunctionPointer = false
+       public virtual Type[] GetFunctionPointerCallingConventions(); 
    }
}
```
### Examples
```cs
// Simple managed function pointer
Type type = typeof(delegate*<int, bool>);
Debug.Assert(type.IsFunctionPointer == true);
Debug.Assert(type.IsUnmanagedFunctionPointer == false);
Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));
Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
```

```cs
// Simple unmanaged function pointer
Type type = typeof(delegate* unmanaged[Cdecl, MemberFunction]<int, bool>);
Debug.Assert(type.IsFunctionPointer == true);
Debug.Assert(type.IsUnmanagedFunctionPointer == true);
Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));

Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
```

## GetModified*Type() methods for `FieldInfo`, `PropertyInfo` and `ParameterInfo`

> If the return type is from the corresponding `FieldInfo.FieldType`, `PropertyInfo.PropertyType` or `ParameterInfo.ParameterType` is a function pointer and has custom modifiers, then a "modified type" is returned which is a wrapper over the type returned from `FieldType \ PropertyType \ Parametertype`. That type will have the property `IsModifierType = true`.

Reviewer 5: ```suggestion
If the return type is from the corresponding `FieldInfo.FieldType`, `PropertyInfo.PropertyType` or `ParameterInfo.ParameterType` is a function pointer and has custom modifiers, then a "modified type" is returned which is a wrapper over the type returned from `FieldType \ PropertyType \ Parametertype`. That type will have the property `IsModifiedType = true`.
```

> If the return type is not a function pointer, or a function pointer without any custom modifiers, then the value from FieldType etc. will be returned which will have `IsModifierType = false`.

Reviewer 5: ```suggestion
If the return type is not a function pointer, or a function pointer without any custom modifiers, then the value from FieldType etc. will be returned which will have `IsModifiedType = false`.
```

 This is a value type, not a reference type like `Pointer`, since newer proposed reflection APIs will not force boxing to occur.

 Like `Pointer` there is no property to obtain the type since it is used in a temporary manner to manually marshal.

```diff
namespace System.Reflection
{
+   public struct FunctionPointer
+   {
+       public static FunctionPointer Box(void* ptr, Type type);
+       public static void* Unbox (FunctionPointer ptr);
+   }
}
```

## Support Type.GetMethod()
If a user tries to retrieve a Method through [`Type.GetMethod()`](https://learn.microsoft.com/dotnet/api/system.type.getmethod) that has function pointer arguments specified by the `Type[]` passed into that method, the current design throws `NotSupportedException` since we don't have a good way to specify the custom modifiers for each parameter in the API (see below -- requires "MethodSignature").

The workaround is to have the caller manually loop through the `MethodInfo`s returned from `Type.GetMethods()` and perform the filtering there.

```diff
    public abstract class Type
    {

>         // This is used with 'Type.GetMethod(..., Type[] types)' to find a method with parameters that

Reviewer 5: Can `typeof(delegate* <void>)` be used with `Type.GetMethod` as well (for runtime types)? Is `ReferenceEquals(Type.MakeFunctionPointerType(typeof(void), Types.EmptyTypes, false) , typeof(delegate* <void>))` going to be `true`? 

I do not see how this API can work well with different type universes, but we do not need to figure this out now that it is in the potential future additions section.

Author: > Can typeof(delegate*<void>) be used with Type.GetMethod as well (for runtime types)?

Good question. IMO it does make sense to support runtime types like this (in addition to those from MakeFunctionPointerType). I'll add a note to the design for this.

> Is ReferenceEquals(Type.MakeFunctionPointerType(typeof(void), Types.EmptyTypes, false) , typeof(delegate* <void>)) going to be true?

No, unless we want to do something more here than just supporting `GetMethod()` lookups. The method comparison logic in `GetMethod()` would match up parameter types\modifiers, so don't need to require `==`, `Equals()` or a `ReferenceEquals()` match. Similar to`Type.MakeGenericSignatureType()`, a type returned from `MakeFunctionPointerType()` will derive from [`SignatureType`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/SignatureType.cs) and isn't expected to be used outside of GetMethod() lookups.

Reviewer 5: > > Is ReferenceEquals(Type.MakeFunctionPointerType(typeof(void), Types.EmptyTypes, false) , typeof(delegate* )) going to be true?

> No, unless we want to do something more here than just supporting GetMethod() lookups.

I think we should do more here if/once we get to this.

All existing `Make*` reflection APIs return the type that you would get by other means (e.g. by inspecting method or field signatures). It would be confusing for `MakeFunctionPointer` to return a function pointer type that is not full-fledged function pointer type.

`MakeGenericSignatureType` is special. It is not possible to get the type returned by `MakeGenericSignatureType` by other means and that's why it does not need to unify with anything.

Author: > I think we should do more here if/once we get to this.

Sure - I'll update the design to suggest the full approach.

        // are function pointers. For a similar reference, see Type.MakeGenericSignatureType().

> +       public static Type MakeFunctionPointerType(Type[] functionPointerArguments, Type returnType, bool isUnmanaged);

Reviewer 5: ```suggestion
+       public static Type MakeFunctionPointerType(Type returnType, Type[] functionPointerArguments, bool isUnmanaged);
```

        // For unmanaged methods, we need to specify the modopts, so something like the proposed MethodSignature below
+       public static Type MakeFunctionPointerType(MethodSignature functionPointerArguments);
        // which is preferred over something like this:
        public static Type MakeFunctionPointerType(

>             Type[] argumentTypes,
>             Type[,] argumentOptionalModifiers,
>             Type[,] argumentRequiredModifiers,
>             Type[] returnType,
>             Type[] returnTypeOptionalModifiers,
>             Type[] returnTypeRequiredModifiers,

Reviewer 5: ```suggestion
            Type returnType,
            Type[]? returnTypeOptionalModifiers,
            Type[]? returnTypeRequiredModifiers,
            Type[] argumentTypes,
            Type[][]? argumentOptionalModifiers,
            Type[][]? argumentRequiredModifiers,
```
To match existing methods like 
https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.methodbuilder.setsignature

Author: Thanks. Updated.

+       public virtual Type[] GetRequiredCustomModifiers();
+       public virtual Type[] GetOptionalCustomModifiers();

        // Throws InvalidOperationException if IsFunctionPointer = false
+       public virtual Type[] GetFunctionPointerCallingConventions(); 
    }
}
```
### Examples
```cs
// Simple managed function pointer
Type type = typeof(delegate*<int, bool>);
Debug.Assert(type.IsFunctionPointer == true);
Debug.Assert(type.IsUnmanagedFunctionPointer == false);
Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));
Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
```

```cs
// Simple unmanaged function pointer
Type type = typeof(delegate* unmanaged[Cdecl]<int, bool>);
Debug.Assert(type.IsFunctionPointer == true);
Debug.Assert(type.IsUnmanagedFunctionPointer == true);
Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));

Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
```

## GetModified*Type() methods for `FieldInfo`, `PropertyInfo` and `ParameterInfo`

> If the return type is from the corresponding `FieldInfo.FieldType`, `PropertyInfo.PropertyType` or `ParameterInfo.ParameterType` is a function pointer and has custom modifiers, then a "modified type" is returned which is a wrapper over the type returned from `FieldType \ PropertyType \ Parametertype`. That type will have the property `IsModifiedType= true`.

Reviewer 5: ```suggestion
If the return type is from the corresponding `FieldInfo.FieldType`, `PropertyInfo.PropertyType` or `ParameterInfo.ParameterType` is a function pointer and has custom modifiers, then a "modified type" is returned which is a wrapper over the type returned from `FieldType \ PropertyType \ Parametertype`. That type will have the property `IsModifiedType = true`.
```
Formatting nit

> If the return type is not a function pointer, or a function pointer without any custom modifiers, then the value from FieldType etc. will be returned which will have `IsModifiedType= false`.

Reviewer 5: ```suggestion
If the return type is not a function pointer, or a function pointer without any custom modifiers, then the value from FieldType etc. will be returned which will have `IsModifiedType = false`.
```

## Modified types
Obtaining all custom modifiers including `ref\out\in` is useful for those who read metadata for varying reasons. The proposal to support that is to expose functionality that returns a "modified type" which, for function pointers, is a wrapper over the unmodified function pointer and overrides `GetFunctionPointerParameterTypes()` to return a potential "modified type" for each parameter. From those modified types, `GetRequiredCustomModifiers()` and `GetOptionalCustomModifiers()` can be used.

> A modified type always start with a "root" function pointer obtained from `FieldInfo.GetModifiedFieldType()`, `PropertyInfo.GetModifiedPropertyType()` and `ParameterInfo.GetModifiedParameterType()`.

Reviewer 5: ```suggestion
A modified type always start with a "root" type obtained from `FieldInfo.GetModifiedFieldType()`, `PropertyInfo.GetModifiedPropertyType()` and `ParameterInfo.GetModifiedParameterType()`.
```
I assume we are not going to special case the modified types for function pointers only. For example, the following will work:

```csharp
var initOnly = typeof(MyStruct).GetMethod("set_MyProperty").ReturnParameter;

Debug.Assert(initOnly.GetModifiedParameterType().GetRequiredCustomModifiers()[0] == typeof(IsExternalInit));

public struct MyStruct
{
    public int MyProperty { get; init; }
}
```

Author: > I assume we are not going to special case the modified types for function pointers only

This was actually a topic I wanted to discuss.

The `init` modifier belongs to the property setter declaration, and not to the `int` type, so this would work today:
```cs
Debug.Assert(initOnly.GetRequiredCustomModifiers()[0] == typeof(IsExternalInit));
```
but I wasn't necessarily assuming `GetModifiedParameterType()` would also include the modifiers (i.e. only function pointer types have embedded modifiers). However, since we are generalizing via "modified type" then it does make sense in the broader "modifier type" usages outside of function pointers. I'll update the design for this then.

var fn = new delegate*<int>[1];
Console.WriteLine(fn is delegate*<int>[]); // true
Console.WriteLine(fn is delegate* unmanaged[Cdecl] <int>[]); // false
```

but unmanaged calling conventions are not part of the type identity:
```cs
object o1 = new delegate* unmanaged[SuppressGCTransition]<int>[1];
Console.WriteLine(o1 is delegate* unmanaged[MemberFunction]<int>[]); // true

object o2 = new delegate* unmanaged[CDecl]<int>[1];
Console.WriteLine(o2 is delegate* unmanaged[StdCall]<int>[]); // true
```

In addition, all non-CallConv* modifiers, such as `ref\in\out\const` are not considered part of the type identity either.

### With modified types
A modified type acts like the underlying type with the exception that it overrides `GetFunctionPointerParameterTypes()` to return the modified parameter types. It has an identity separate than the underlying type.

```cs
public class Holder
{
    public unsafe delegate* unmanaged[Cdecl, MemberFunction]<int> _field;
    public unsafe delegate* unmanaged[Cdecl, MemberFunction]<int> _field2;
}

Type modifiedType = typeof(Holder).GetField("_field").GetModifiedFieldType();
Console.WriteLine(modifiedType.IsFunctionPointer); // true

> Console.WriteLine(modifiedType.UnderlyingSystemType != null); // true (unmodified function pointer)

Reviewer 5: ```suggestion
Console.WriteLine(!ReferenceEquals(modifiedType.UnderlyingSystemType,  modifiedType)); // true (unmodified function pointer)
```
UnderlyingSystemType is never null.

```cs
    string s = typeof(Action<int>).Name; // "Action`1
```

### `Namespace` property
Returns `null`.

A standard pointer returns the namespace of the referenced type, but we don't have that with function pointers since they are declared inline.

### `IsPointer` property
Returns `false`. Per past design discussions, this should be `false` since function pointers and standard pointers are not the same, and this also is non-breaking for existing code that needs to determine if a type is a standard pointer type.

### `IsValueType` and `IsClass` properties
Use the same behavior as standard pointers which is `IsValueType == false` and `IsClass == true`.

The type hierarchy is covered somewhat [here](https://learn.microsoft.com/previous-versions/visualstudio/visual-studio-2008/2hf02550(v=vs.90)) and in ECMA\CLI: 
- Type
  - Value types
    - Built-in value types
    - User-defined value types
    - Enumerations
  - Reference types
    - **Pointer types**
    - Interface types
    - Self-describing types
      - Arrays
      - Class types
        - User-defined classes
        - Boxed value types
        - Delegates

> ### `Type? UnderlyingSystemType { get; }`

Reviewer 5: (I am not commenting on all olaces in the last edit that assumed that `UnderlyingSystemType` returns null.)

> ### `Type? UnderlyingSystemType { get; }`

Author: Updated.

In CLI metadata, only fields, properties, local variables, method parameters and function pointer parameters allow custom modifiers. A type, method or function pointer itself cannot have custom modifiers; function pointers use the "return parameter" to contain its modopts (optional custom modifiers) used for calling conventions. Thus, the new `GetModified*Type()` methods could be called `GetFunctionPointerModified*Type()` instead. However, the shorter version is used here just in case this no longer holds to just function pointers.

For additional information on encoding, see ECMA-335 `II.23.2 Blobs and signatures`. Note that in C# a function pointer is emitted as an ECMA "StandaloneMethodSig". There are several other types of custom modifiers besides the calling convention ones. For example, in C#, a field is also a "signature" and can have a `volatile` modifier which is emitted as a custom modifier and will be returned via `FieldInfo.GetModifiedFieldType().GetRequiredCustomModifiers()` as the `System.Runtime.CompilerServices.IsVolatile` type. In C++\CLI there are additional modifiers including `System.Runtime.CompilerServices.IsConst`.

> Supporting modified types correctly means also supporting "nested types" including

Reviewer 5: Nested types are something else - e.g. see `Type.GetNestedTypes()` or look for `Nested type` in ECMA-335.

I think these are called ""parametrized types".

Author: Yep thanks.

- Arrays
```cs
    delegate*<int>[] a_0 = null!;
    delegate*<int>[][] a_1 = null!;
    delegate*<int>[,] a_2 = null!;
```
- Pointers
```cs
    delegate*<int>* p_0 = null;
    delegate*<int>** p_1 = null;
    delegate*<int>*** p_2 = null;
```
- Parameters of a parent function pointer
```cs
    delegate*<delegate*<int>> fp_0 = null;
```
- Combinations of the above
```cs
    // Arrays + pointers
    delegate*<int>*[] ap_0 = null!;
    delegate*<int>**[] ap_1 = null!;
    delegate*<int>*[][] ap_2 = null!;

    // Parent fcnptr + pointers
    delegate*<delegate*<int>*> fp_1 = null;

    // Parent fcnptr + arrays
    delegate*<delegate*<int>[]> fp_2 = null;

    // Parent fcnptr + arrays + pointers
    delegate*<delegate*<int>*[]> fp_3 = null;
```

> Note that a function pointer type cannot currently be used as a generic parameter type, so that limits this scope a bit.

Reviewer 5: You can have arrays of function pointers as a generic parameter. For example, `new List<delegate*<int>[]>();` compiles fine. Is that going to work as expected?

Author: Yes.

Consider the following where `ref\out\in` all map to a reference (`&`) by the runtime:
```cs
class MethodHolder
{
    public static void M1(in int i) { }
    public static void M2(ref int i) { }
    public static void M3(out int i) { i = 42; }
}

Type t1 = typeof(MethodHolder).GetMethod("M1").GetParameters()[0].ParameterType;
Type t2 = typeof(MethodHolder).GetMethod("M2").GetParameters()[0].ParameterType;
Type t3 = typeof(MethodHolder).GetMethod("M3").GetParameters()[0].ParameterType;

Debug.Assert(t1 == t2 && t2 == t3); // True today; all are "int reference type" (int&)
```

Also consider [ILGenerator.EmitCalli()](https://learn.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli(system-reflection-emit-opcode-system-runtime-interopservices-callingconvention-system-type-system-type())) as a reference implementation which does not support modifiers:
```cs
void EmitCalli(
    System.Reflection.Emit.OpCode opcode,
    System.Runtime.InteropServices.CallingConvention unmanagedCallConv,
    Type? returnType,
    Type[]? parameterTypes)
```

> So custom modifiers, except for calling conventions, are not exposed or used by the native runtime. They are not necessary for the key scenario of passing or invoking a function pointer and would also interfere with equality and castability. If function pointers expose all custom modifiers, such as `ref\in` and the C++/CLI `const`, then equality and castability will be based on exact matches to that which is considered overly restrictive for runtime use and may break C++/CLI usage.

Author: I've updated this to focus on function pointer types maintained by the runtime, and not other areas that need the metadata (which are not obtained via the function pointer type \ TypeDesc).

}
```

## Support Type.GetMethod()

> If a user tries to retrieve a Method through [`Type.GetMethod()`](https://learn.microsoft.com/dotnet/api/system.type.getmethod) that has function pointer arguments specified by the `Type[]` passed into that method, the current design throws `NotSupportedException` since we don't have a good way to specify the custom modifiers for each parameter in the API (see below -- requires "MethodSignature").

Reviewer 3: FWIW, C# doesn't actually allow things to overload if the only difference is in the modifiers that shouldn't matter.

```cs
unsafe class Foo
{
    // Fine
    void A(delegate* unmanaged[Stdcall]<int, void> x) => throw null;
    void A(delegate* unmanaged[Cdecl]<int, void> x) => throw null;

    // Doesn't compile
    void B(delegate*<ref int, void> x) => throw null;
    void B(delegate*<out int, void> x) => throw null;
}
```

So if we did care about the callconv but nothing else, it would be okay for C#. It would still not be okay for other languages, but that's how it is with modifiers right now in general anyway - we ignore them in many positions.

The more I think about it, the more it feels like the `out`/`ref` bits on function pointers should have been encoded as a custom attribute, not as a modifier - same way C# encodes nullability information. The custom modifier throws off a lot of things. If we could go back on that, it would probably help.

Author: > The more I think about it, the more it feels like the out/ref bits on function pointers should have been encoded as a custom attribute, not as a modifier - same way C# encodes nullability information. The custom modifier throws off a lot of things. If we could go back on that, it would probably help.

Since function pointers can't have attributes or pseudo-attributes like `[in]` I suppose it was the most straightforward to use modopts. However, I wonder what it would look like if we could go back on ECMA-335 instead and change the somewhat rigid spec of how attributes, pseudo-attributes and modifiers each can\can't be applied to the various typedecls\refs\signatures\parameters.

Reviewer 3: > Since function pointers can't have attributes or pseudo-attributes like `[in]` I suppose it was the most straightforward to use modopts.

We didn't use modopts to encode nullability:

```csharp
class Program
{
    static void Method1(List<object?> param) { }
    static void Method2(List<object> param) { }
}
```

This has the same problem - how C# can keep of the track of the question mark in `Method1`? We could put a modopt on the `object` part, but that affects overload resolution. Instead this is encoded in a custom attribute on the method. An encoding was invented to have this information in a sidecar, no matter how nested the question mark needs to be.

The `ref`/`out`/`in`/`readonly` etc. bits should have gone into a similar sidecar so that they don't mess with overload resolution.
